File: go/internal/queue/codec.go
package queue

import (
	"fmt"
	"strconv"

	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"

	codev1 "coda/go/internal/genproto/coda/v1"
)

// marshalOpts is the one protojson configuration used on the wire.
// EmitUnpopulated is off: a zero-valued field means "absent" identically on
// both sides, and omitting them keeps `XRANGE` output readable, which is
// the whole reason ADR-0004 chose protojson over binary protobuf.
var marshalOpts = protojson.MarshalOptions{}

// unmarshalOpts tolerates fields this build does not know about, so a
// worker running a newer minor contract does not hard-fail a message it
// could otherwise process. Genuinely breaking changes are caught by the
// SchemaVersion check instead (§2.2), which is an explicit, loud FATAL
// rather than a silent misread.
var unmarshalOpts = protojson.UnmarshalOptions{DiscardUnknown: true}

// fields renders a proto message into the flat Redis stream entry shape.
// The duplicated scalars (job_id, stage, attempt, idempotency_key,
// trace_id) are denormalised out of the payload on purpose: they make a
// stream entry greppable and let the reaper decide what to do with a
// reclaimed message without unmarshalling it.
func fields(kind string, m proto.Message, extra map[string]string) (map[string]any, error) {
	payload, err := marshalOpts.Marshal(m)
	if err != nil {
		return nil, fmt.Errorf("queue: marshal %s: %w", kind, err)
	}
	out := map[string]any{
		FieldKind:          kind,
		FieldPayload:       string(payload),
		FieldSchemaVersion: strconv.FormatUint(uint64(SchemaVersion), 10),
	}
	for k, v := range extra {
		if v != "" {
			out[k] = v
		}
	}
	return out, nil
}

func envelopeFields(env *codev1.StageEnvelope) (map[string]any, error) {
	return fields(KindStageEnvelope, env, map[string]string{
		FieldJobID:          env.GetJobId(),
		FieldStage:          StageName(env.GetStage()),
		FieldAttempt:        strconv.FormatUint(uint64(env.GetAttempt()), 10),
		FieldIdempotencyKey: env.GetIdempotencyKey(),
		FieldTraceID:        env.GetTraceId(),
	})
}

func resultFields(res *codev1.StageResult) (map[string]any, error) {
	return fields(KindStageResult, res, map[string]string{
		FieldJobID:          res.GetJobId(),
		FieldStage:          StageName(res.GetStage()),
		FieldAttempt:        strconv.FormatUint(uint64(res.GetAttempt()), 10),
		FieldIdempotencyKey: res.GetIdempotencyKey(),
		FieldTraceID:        res.GetTraceId(),
	})
}

func heartbeatFields(hb *codev1.StageHeartbeat) (map[string]any, error) {
	return fields(KindStageHeartbeat, hb, map[string]string{
		FieldJobID:   hb.GetJobId(),
		FieldStage:   StageName(hb.GetStage()),
		FieldAttempt: strconv.FormatUint(uint64(hb.GetAttempt()), 10),
		FieldTraceID: hb.GetTraceId(),
	})
}

func deadLetterFields(dl *codev1.DeadLetter) (map[string]any, error) {
	return fields(KindDeadLetter, dl, map[string]string{
		FieldJobID:   dl.GetJobId(),
		FieldStage:   StageName(dl.GetStage()),
		FieldTraceID: dl.GetOriginalEnvelope().GetTraceId(),
	})
}

// ErrSchemaUnsupported is §2.2's rule made a value: a message whose
// schema_version this build does not implement is rejected, never guessed
// at. The orchestrator turns it into a FATAL / SCHEMA_UNSUPPORTED
// dead-letter rather than a retry, since retrying cannot make an
// unimplemented contract implementable.
var ErrSchemaUnsupported = fmt.Errorf("queue: unsupported schema_version")

// decode unmarshals a message payload into m after checking its kind and
// schema version.
func decode(msg Message, wantKind string, m proto.Message) error {
	if got := msg.Fields[FieldKind]; got != "" && got != wantKind {
		return fmt.Errorf("queue: message %s on %s is a %s, want %s", msg.ID, msg.Stream, got, wantKind)
	}
	if raw := msg.Fields[FieldSchemaVersion]; raw != "" {
		v, err := strconv.ParseUint(raw, 10, 32)
		if err != nil {
			return fmt.Errorf("queue: message %s has unparseable schema_version %q: %w", msg.ID, raw, err)
		}
		if uint32(v) != SchemaVersion {
			return fmt.Errorf("%w: message %s declares %d, this build implements %d", ErrSchemaUnsupported, msg.ID, v, SchemaVersion)
		}
	}
	payload := msg.Fields[FieldPayload]
	if payload == "" {
		return fmt.Errorf("queue: message %s on %s has no %s field", msg.ID, msg.Stream, FieldPayload)
	}
	if err := unmarshalOpts.Unmarshal([]byte(payload), m); err != nil {
		return fmt.Errorf("queue: unmarshal %s from message %s: %w", wantKind, msg.ID, err)
	}
	return nil
}

// DecodeEnvelope reads a StageEnvelope out of a consumed message.
func DecodeEnvelope(msg Message) (*codev1.StageEnvelope, error) {
	env := &codev1.StageEnvelope{}
	return env, decode(msg, KindStageEnvelope, env)
}

// DecodeResult reads a StageResult out of a consumed message.
func DecodeResult(msg Message) (*codev1.StageResult, error) {
	res := &codev1.StageResult{}
	return res, decode(msg, KindStageResult, res)
}

// DecodeHeartbeat reads a StageHeartbeat out of a consumed message.
func DecodeHeartbeat(msg Message) (*codev1.StageHeartbeat, error) {
	hb := &codev1.StageHeartbeat{}
	return hb, decode(msg, KindStageHeartbeat, hb)
}

// DecodeDeadLetter reads a DeadLetter out of a consumed message — used by
// the DLQ alerting job and by operator replay tooling.
func DecodeDeadLetter(msg Message) (*codev1.DeadLetter, error) {
	dl := &codev1.DeadLetter{}
	return dl, decode(msg, KindDeadLetter, dl)
}

File: python/shared/src/coda_worker_sdk/codec.py
"""protojson encode/decode against the generated coda.v1 contract classes.

Mirrors go/internal/queue/codec.go field-for-field: the same denormalised
scalars (job_id, stage, attempt, idempotency_key, trace_id) sit alongside the
protojson payload so a stream entry stays greppable via `XRANGE`, and
schema_version is checked on every decode (architecture.md §2.2/§3.1).
"""

from __future__ import annotations

from typing import TypeVar

from google.protobuf import json_format
from google.protobuf.message import Message

from coda.v1 import common_pb2, envelope_pb2
from coda_worker_sdk.streams import (
    FIELD_ATTEMPT,
    FIELD_IDEMPOTENCY_KEY,
    FIELD_JOB_ID,
    FIELD_KIND,
    FIELD_PAYLOAD,
    FIELD_SCHEMA_VERSION,
    FIELD_STAGE,
    FIELD_TRACE_ID,
    KIND_DEAD_LETTER,
    KIND_STAGE_ENVELOPE,
    KIND_STAGE_HEARTBEAT,
    KIND_STAGE_RESULT,
    SCHEMA_VERSION,
    stage_name,
)

M = TypeVar("M", bound=Message)


class SchemaUnsupportedError(Exception):
    """A message declared a schema_version this build does not implement
    (architecture.md §2.2). The caller must classify this FATAL, never retry
    it — retrying cannot make an unimplemented contract implementable."""


def encode_fields(kind: str, message: Message, extra: dict[str, str]) -> dict[str, str]:
    """Render a proto message into the flat Redis stream entry shape used by
    every stream in this system. `extra` supplies the denormalised scalars;
    empty values are omitted, matching the Go side's `fields()`.
    """
    payload = json_format.MessageToJson(message, preserving_proto_field_name=True, indent=None)
    out: dict[str, str] = {
        FIELD_KIND: kind,
        FIELD_PAYLOAD: payload,
        FIELD_SCHEMA_VERSION: str(SCHEMA_VERSION),
    }
    for k, v in extra.items():
        if v:
            out[k] = v
    return out


def result_fields(res: envelope_pb2.StageResult) -> dict[str, str]:
    return encode_fields(
        KIND_STAGE_RESULT,
        res,
        {
            FIELD_JOB_ID: res.job_id,
            FIELD_STAGE: stage_name(res.stage),
            FIELD_ATTEMPT: str(res.attempt),
            FIELD_IDEMPOTENCY_KEY: res.idempotency_key,
            FIELD_TRACE_ID: res.trace_id,
        },
    )


def heartbeat_fields(hb: envelope_pb2.StageHeartbeat) -> dict[str, str]:
    return encode_fields(
        KIND_STAGE_HEARTBEAT,
        hb,
        {
            FIELD_JOB_ID: hb.job_id,
            FIELD_STAGE: stage_name(hb.stage),
            FIELD_ATTEMPT: str(hb.attempt),
            FIELD_TRACE_ID: hb.trace_id,
        },
    )


def decode(fields: dict[str, str], want_kind: str, message: M) -> M:
    """Unmarshal a message payload after checking its kind and schema
    version. Mirrors go/internal/queue/codec.go's `decode`.
    """
    got_kind = fields.get(FIELD_KIND, "")
    if got_kind and got_kind != want_kind:
        raise ValueError(f"coda_worker_sdk: message is a {got_kind}, want {want_kind}")

    raw_version = fields.get(FIELD_SCHEMA_VERSION, "")
    if raw_version:
        try:
            version = int(raw_version)
        except ValueError as exc:
            raise ValueError(
                f"coda_worker_sdk: unparseable schema_version {raw_version!r}"
            ) from exc
        if version != SCHEMA_VERSION:
            raise SchemaUnsupportedError(
                f"message declares schema_version {version}, this build implements {SCHEMA_VERSION}"
            )

    payload = fields.get(FIELD_PAYLOAD, "")
    if not payload:
        raise ValueError(f"coda_worker_sdk: message has no {FIELD_PAYLOAD!r} field")
    json_format.Parse(payload, message, ignore_unknown_fields=True)
    return message


def decode_envelope(fields: dict[str, str]) -> envelope_pb2.StageEnvelope:
    return decode(fields, KIND_STAGE_ENVELOPE, envelope_pb2.StageEnvelope())


def decode_dead_letter(fields: dict[str, str]) -> envelope_pb2.DeadLetter:
    return decode(fields, KIND_DEAD_LETTER, envelope_pb2.DeadLetter())


__all__ = [
    "SchemaUnsupportedError",
    "encode_fields",
    "result_fields",
    "heartbeat_fields",
    "decode",
    "decode_envelope",
    "decode_dead_letter",
    "common_pb2",
]
